### PR TITLE
Refactor Plaquette hierarchy

### PR DESCRIPTION
**Pull request by nelimee**
This PR try to address #136.

### CODE REVIEW

**Comment by nelimee** [comment]
Note: removing the `PlaquetteList` class made a few notebooks fail. `PlaquetteList` has always been too specialised to be in the main library in my opinion (only useful for simple memory experiments), so I will change the notebooks accordingly.

**Comment by nelimee** [comment]
The notebook [`move_qubit_along_a_line.ipynb`](https://github.com/QCHackers/tqec/blob/main/notebooks/move_qubit_along_a_line.ipynb) should still be updated to the new plaquette hierarchy, but the other notebooks are working as expected.

**Comment by smburdick** [comment]
@nelimee Would you like my review on this?

**Comment by nelimee** [comment]
I am still missing the rewriting of one notebook, I am having a hard time getting the observable correct, but everything else should be near final so if you have the time, your review is more than welcome 😊
The more opinions we have, the better it is!

**Comment by nelimee** [comment]
With this last commit, I did everything I planned with this PR. I am ready to discuss/make changes according to your reviews :)

**Comment by smburdick** [comment]
> With this last commit, I did everything I planned with this PR. I am ready to discuss/make changes according to your reviews :)

I'll see what I can do. I would like to familiarize myself with the backend more.

**Comment by nelimee** [comment]
> I'll see what I can do. I would like to familiarize myself with the backend more.

Feel free to ask me questions if you have any, we can also schedule a meeting like last time :)

**Comment by nelimee** [comment]
I will also, either this week-end or at the beginning of next week, refactor the different measurement plaquettes. All the `{XX,ZZ,XXXX,ZZZZ}FinalMeasurementPlaquette` will be replaced by two plaquettes `MeasurementRoundedPlaquette` and `MeasurementSquarePlaquette` in a new `tqec.plaquettes.library.memory` module. 

With the exception of that change (that is simply a code factorisation step), this PR should still be considered finalised and ready to be reviewed.

**Comment by inmzhang** [comment]
I've gone through most of the refactoring changes in the associated PR #100 and initially had few opinions. I will revisit this tomorrow for a closer look.

File: src/tqec/detectors/operation.py [hunk 1]
```diff
@@ -59,6 +59,9 @@ def qubits(self) -> tuple[cirq.Qid, ...]:
     def with_qubits(self, *new_qubits: cirq.Qid) -> "ShiftCoords":
         return self
 
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({self._shifts})"
```
Comment (inmzhang) [inline src/tqec/detectors/operation.py:63]:
```suggestion
    def __repr__(self) -> str:
        return f"{self.__class__.__name__}{self._shifts}"
```

File: src/tqec/plaquette/library/initialisation.py
Comment (inmzhang) [inline src/tqec/plaquette/library/initialisation.py:1]: All the initialization/measurement plaquettes for X/Z basis and square/rounded variants are almost the same. I think we can refactor it further by extracting the circuit construction process.  Currently, the difference between the square/rounded variants is only the way we get the data qubits. If we can make `get_data_qubits_cirq(self)` as well as `get_syndrome_qubits_cirq(self)` as an abstract method in the `Plaquette` level, we can eliminate some code repetition in these cases.

File: src/tqec/plaquette/plaquette.py [hunk 2]
```diff
@@ -148,19 +164,21 @@ def __init__(
...
-        PlaquetteQubit(Position(0, 0)),
-        PlaquetteQubit(Position(2, 0)),
-        PlaquetteQubit(Position(0, 2)),
-        PlaquetteQubit(Position(2, 2)),
-    ]
-    _syndrome_qubits = [PlaquetteQubit(Position(1, 1))]
+    @staticmethod
+    def _get_data_qubit_side_from_plaquette_orientation(
+        orientation: PlaquetteOrientation,
+    ) -> PlaquetteSide:
```
Comment (inmzhang) [inline src/tqec/plaquette/plaquette.py:173]: This relationship can be defined within the `PlaquetteOrientation` or `PlaquetteSide` enum.

File: src/tqec/plaquette/library/initialisation.py
Comment (nelimee) [reply src/tqec/plaquette/library/initialisation.py:1]: The problem of making the `get_data_qubits` and `get_data_qubits_cirq` methods non-static is that, most of the time, we need them before actually calling `__init__` on the `Plaquette` instance. For example in
https://github.com/QCHackers/tqec/blob/3df4a0588ef9aacd27e553af06b1af6ab38d40d9/src/tqec/plaquette/library/zz.py#L11-L33
we need to access the data qubits to create the circuit, before actually calling the `RoundedPlaquette.__init__` (which is exactly `super().__init__` here) method.

That could be solved by first creating the `Plaquette` instance with an empty circuit and then change that circuit, but this is not really desirable in my opinion.

Any idea how we could proceed?

Comment (inmzhang) [reply src/tqec/plaquette/library/initialisation.py:1]: > The problem of making the get_data_qubits and get_data_qubits_cirq methods non-static is that, most of the time, we need them before actually calling __init__ on the Plaquette instance. 

That makes sense to me. Anyway, we can at least refactor some of the circuit construction code, take syndrome measurement for example:
```python
def make_syndrom_measurement_circuit(
    syndrome_qubit: cirq.GridQubit, 
    ordered_data_qubits: list[cirq.GridQubit],
    detector: cirq.Operation | None,
    x_basis: bool = False
) -> cirq.Circuit:
    ...
```
Then we can use it in all XXXX/ZZZZ/XX/ZZ and Square/Rounded plaquettes.

**Comment by nelimee** [comment]
Following the last commit, I have a few conceptual questions. 

First, currently, two "features" of a `Plaquette` may be passed down by inheritance: 
1. its shape: is it square or rounded
2. its effect: is it a `XX` syndrome measurement, `XXXX`, `ZZ`, `ZZZZ`?

Initially, the `RoundedPlaquette` and `SquarePlaquette` have been introduced to have a common definition **and order** of the data qubits the plaquette should operate on, as well as to have helper methods to recover the correct data qubits in the case of `RoundedPlaquette` instances. 

Now, I really think that it is more meaningful to have `XXMemoryPlaquette` (that should probably be renamed `XXSyndromeMeasurementPlaquette`) inherit from a `PauliMemoryPlaquette` (should be renamed too if introduced) than from a `RoundedPlaquette`.

Any opinion on that?

**Comment by inmzhang** [comment]
> Now, I really think that it is more meaningful to have XXMemoryPlaquette (that should probably be renamed XXSyndromeMeasurementPlaquette) inherit from a PauliMemoryPlaquette (should be renamed too if introduced) than from a RoundedPlaquette.

LGTM. This kind of syndrome measurements should have the same structure at least at the stabilizer level. One comment which is not quite related to this PR, on the native gateset we used in the circuit implementation: While you have chosen the CNOT gate, I can think of another two ways to implement the controlled-pauli gates:

1. Use A control B gates, where A/B are control and target basis for the gate. This has the benefits of completely removing the basis change gates and is natively supported by `stim`. In the simple X/Z plaquette setting, this might be overkilling. But it can be useful in more complicated pauli syndrome measurements. However, these gates seem nontrivial in `cirq` and will need to be transpiled to some experimental native gateset if we want to simulate for experiment setups.
2. Use CZ with basis-change single qubit gates. This has the benefits of being symmetry and is native to superconducting qubits(and some other systems). But this can complicate the simple X/Z syndrome circuits a little bit.

I'm still fine with the CNOT, which can be transpiled to CZ gateset in the post circuit transformation.

**Review by inmzhang — APPROVED** [review]

**Comment by nelimee** [comment]
I re-tested and re-read everything, this PR was a little bit too large, thanks @inmzhang for reviewing! I'll adapt #130 and merge as soon as possible.

**Pull request merged**